### PR TITLE
Make bezier handle type a property of keyframes, update UI

### PR DESCRIPTION
**Pull request by NathanLovato**
- Remove unused code related to old close icon (now replaced with a button)
- Add bezier handle options to right-click menu
- Provide optional scale for balanced handle move
- Remove mirror handle mode, only keep balanced. The reason is it's not one you're likely to use more than rarely. I've only seen this in Godot when it comes to animation. Balanced keeps things aligned so it does the job well.
- Update animation reference

This PR implements https://github.com/godotengine/godot-proposals/issues/3236

It makes the bezier handle mode a property of each bezier track keyframe, allowing to retain some behavior on the keyframes. It opens the path to having auto handles as explained by @TokageItLab here: https://github.com/godotengine/godot-proposals/issues/3141#issuecomment-922247544

Co-authored-by: Francois Belair <razoric480@gmail.com>
Co-authored-by: Gilles Roudière <gilles.roudiere@gmail.com>

### CODE REVIEW

**Comment by NathanLovato** [comment]
The PR has two bugs left to address and I'd appreciate some help on those, mostly due to having very limited time to work on this (preparing a Kickstarter campaign).

1. The handle mode appears in the inspector but when changing it, it doesn't get saved.

I clearly didn't get the setter right but I'm not sure what's wrong. Here's a video clip:

https://user-images.githubusercontent.com/12694995/135639414-3c8b311d-e298-4e28-b500-dac5bcd56709.mp4

2. The values applied to handles are off.

Before the PR the code to update the handles' position was inside the bezier editor. With this PR the code moves to the Animation class because we need to update the right handle when moving the left one and vice-versa.

There was code like this in the bezier animation editor to update the opposite of the edited handle in balanced mode ("mirrors" the direction of the edited handle but preserves the opposite handle's length):

```cpp
Vector2 scale = Vector2(timeline->get_zoom_scale(), v_zoom);
moving_handle_right = (-(moving_handle_left * scale).normalized() * (moving_handle_right * scale).length()) / scale;
```

Now, the thing is, the coordinates used in the bezier editor seemed to be in different units on the X and Y axes: small on the X-axis, and larger on the Y-axis.

Now, because the calculation on `moving_handle_right` needs to happen in the Animation class now, I don't have access to the scale to get the editor scale and normalized handle correctly when calculating the opposite handle:

```cpp
void Animation::bezier_track_set_key_out_handle(int p_track, int p_index, const Vector2 &p_handle) {
//...
	Vector2 out_handle = p_handle;
	if (out_handle.x < 0) {
		out_handle.x = 0;
	}
	bt->values.write[p_index].value.out_handle = out_handle;

	if (bezier_track_get_key_handle_mode(p_track, p_index) == HANDLE_MODE_BALANCED) {
		const Vector2 in_handle = -out_handle.normalized() * bt->values[p_index].value.in_handle.length();
		bt->values.write[p_index].value.in_handle = in_handle;
	}
```

I'm not sure how to best address that.

Right now it causes keyframes to behave like this:

https://user-images.githubusercontent.com/12694995/135639663-05d5c7da-2619-40b3-a567-8af1514b216a.mp4

**Comment by bruvzg** [comment]
> The handle mode appears in the inspector but when changing it, it doesn't get saved.
I clearly didn't get the setter right but I'm not sure what's wrong. Here's a video clip:


Serialization seems to be done in the `_set`/`_get` methods, also enum and function bindings are missing. Something along these lines should fix it (I have not tested it):

```diff
diff --git a/scene/resources/animation.cpp b/scene/resources/animation.cpp
index caff93903b..0b304653ef 100644
--- a/scene/resources/animation.cpp
+++ b/scene/resources/animation.cpp
@@ -200,7 +200,7 @@ bool Animation::_set(const StringName &p_name, const Variant &p_value) {
 				Vector<real_t> times = d["times"];
 				Vector<real_t> values = d["points"];
 
-				ERR_FAIL_COND_V(times.size() * 5 != values.size(), false);
+				ERR_FAIL_COND_V(times.size() * 6 != values.size(), false);
 
 				if (times.size()) {
 					int valcount = times.size();
@@ -213,11 +213,12 @@ bool Animation::_set(const StringName &p_name, const Variant &p_value) {
 					for (int i = 0; i < valcount; i++) {
 						bt->values.write[i].time = rt[i];
 						bt->values.write[i].transition = 0; //unused in bezier
-						bt->values.write[i].value.value = rv[i * 5 + 0];
-						bt->values.write[i].value.in_handle.x = rv[i * 5 + 1];
-						bt->values.write[i].value.in_handle.y = rv[i * 5 + 2];
-						bt->values.write[i].value.out_handle.x = rv[i * 5 + 3];
-						bt->values.write[i].value.out_handle.y = rv[i * 5 + 4];
+						bt->values.write[i].value.value = rv[i * 6 + 0];
+						bt->values.write[i].value.in_handle.x = rv[i * 6 + 1];
+						bt->values.write[i].value.in_handle.y = rv[i * 6 + 2];
+						bt->values.write[i].value.out_handle.x = rv[i * 6 + 3];
+						bt->values.write[i].value.out_handle.y = rv[i * 6 + 4];
+						bt->values.write[i].value.handle_mode = rv[i * 6 + 5];
 					}
 				}
 
@@ -470,7 +471,7 @@ bool Animation::_get(const StringName &p_name, Variant &r_ret) const {
 				int kk = bt->values.size();
 
 				key_times.resize(kk);
-				key_points.resize(kk * 5);
+				key_points.resize(kk * 6);
 
 				real_t *wti = key_times.ptrw();
 				real_t *wpo = key_points.ptrw();
@@ -481,11 +482,12 @@ bool Animation::_get(const StringName &p_name, Variant &r_ret) const {
 
 				for (int i = 0; i < kk; i++) {
 					wti[idx] = vls[i].time;
-					wpo[idx * 5 + 0] = vls[i].value.value;
-					wpo[idx * 5 + 1] = vls[i].value.in_handle.x;
-					wpo[idx * 5 + 2] = vls[i].value.in_handle.y;
-					wpo[idx * 5 + 3] = vls[i].value.out_handle.x;
-					wpo[idx * 5 + 4] = vls[i].value.out_handle.y;
+					wpo[idx * 6 + 0] = vls[i].value.value;
+					wpo[idx * 6 + 1] = vls[i].value.in_handle.x;
+					wpo[idx * 6 + 2] = vls[i].value.in_handle.y;
+					wpo[idx * 6 + 3] = vls[i].value.out_handle.x;
+					wpo[idx * 6 + 4] = vls[i].value.out_handle.y;
+					wpo[idx * 6 + 5] = vls[i].value.handle_mode;
 					idx++;
 				}
 
@@ -2709,6 +2711,9 @@ void Animation::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("audio_track_get_key_start_offset", "track_idx", "key_idx"), &Animation::audio_track_get_key_start_offset);
 	ClassDB::bind_method(D_METHOD("audio_track_get_key_end_offset", "track_idx", "key_idx"), &Animation::audio_track_get_key_end_offset);
 
+    ClassDB::bind_method(D_METHOD("bezier_track_set_key_handle_mode", "track", "index", "key_handle_mode"), &Animation::bezier_track_set_key_handle_mode);
+    ClassDB::bind_method(D_METHOD("bezier_track_get_key_handle_mode", "track", "index"), &Animation::bezier_track_get_key_handle_mode);
+
 	ClassDB::bind_method(D_METHOD("animation_track_insert_key", "track_idx", "time", "animation"), &Animation::animation_track_insert_key);
 	ClassDB::bind_method(D_METHOD("animation_track_set_key_animation", "track_idx", "key_idx", "animation"), &Animation::animation_track_set_key_animation);
 	ClassDB::bind_method(D_METHOD("animation_track_get_key_animation", "track_idx", "key_idx"), &Animation::animation_track_get_key_animation);
@@ -2746,6 +2751,10 @@ void Animation::_bind_methods() {
 	BIND_ENUM_CONSTANT(UPDATE_DISCRETE);
 	BIND_ENUM_CONSTANT(UPDATE_TRIGGER);
 	BIND_ENUM_CONSTANT(UPDATE_CAPTURE);
+
+    BIND_ENUM_CONSTANT(HANDLE_MODE_FREE);
+    BIND_ENUM_CONSTANT(HANDLE_MODE_BALANCED);
+    BIND_ENUM_CONSTANT(HANDLE_MODE_MIRROR);
 }
 
 void Animation::clear() {
diff --git a/scene/resources/animation.h b/scene/resources/animation.h
index 80e69c10ee..39afc1fa6d 100644
--- a/scene/resources/animation.h
+++ b/scene/resources/animation.h
@@ -367,5 +367,7 @@ public:
 VARIANT_ENUM_CAST(Animation::TrackType);
 VARIANT_ENUM_CAST(Animation::InterpolationType);
 VARIANT_ENUM_CAST(Animation::UpdateMode);
+VARIANT_ENUM_CAST(Animation::HandleMode);
+
 
 #endif

```

**Comment by NathanLovato** [comment]
Thanks much @bruvzg, I didn't check the Animation class for serialization. I thought it might happen in the editor classes.

**Comment by Razoric480** [comment]
Fixed issue 1.

For number 2, we need the animation to handle the balancing of the handles or else it defeats the purpose of putting the handles there. So the only real solution is to provide the scale when calling `bezier_track_set_key_in/out_handle`.

We can make this `p_balanced_scale` parameter optional and default it to `Vector2(1, 1)` to prevent breaking. For most purposes, using it in GDScript would not have the concept of being zoomed in or out unless they're making a custom editor (then it's there for them) so the (1, 1) default sounds about right.

More importantly we can bring it into `animation.cpp` using the original 4.0's (get_zoom_scale(), v_zoom) scale.

https://user-images.githubusercontent.com/2823241/135694736-29c6475b-df0f-4806-8728-0542e6dcc72c.mp4

It is an API change, however, and would need it be added to the docs to indicate its purpose and how it only comes into effect when the key is balanced. But it's the simplest solution by far, and we were already changing the docs by adding the handle enum to the Animation resource.

**Comment by NathanLovato** [comment]
Thanks much @Razoric480 I'll add the docs now to complete checks

**Comment by Razoric480** [comment]
Might need to change the D_METHOD call in Animation.cpp's bind_method as well.

**Comment by NathanLovato** [comment]
@Razoric480 I just did that and finished removing the mirror handle mode.

After testing the new changes I found two bugs:

1. changing a handle's mode from FREE to BALANCED in the inspector can cause the handles to get very long or very short. That's the issue with not having the scale I was mentioning earlier.
2. We can't edit the values of the in and out handles anymore directly in the inspector.

**Comment by NathanLovato** [comment]
One thing I'm also seeing is the handles start in FREE mode, they should probably start in BALANCED. Gonna change this right away.

**Comment by NathanLovato** [comment]
There's one last issue left, and this one might require a breaking change.

Changing a handle's mode from FREE to BALANCED in the inspector can cause the handles to get very long or very short. That's the issue with not having the scale I was mentioning earlier.

The underlying issue is that the handles' coordinates are very different on the X and Y axes: X has small values (in seconds?) while Y has large values (seems in pixels). So when changing the handle mode to balanced, we need to align the left handle with the right one, but because of the disparity between X and Y values, this gives unexpected results.

**Comment by Razoric480** [comment]
The scale for handles, after investigation, is actually in `(+/- amount of time offset from key, +/- number of units offset from key)`, which simplifies things a lot as we no longer need to think of zooming or the size of the bezier editor's control.

**Comment by NathanLovato** [comment]
It's much better but not quite there yet. In balanced mode, the opposite handle should always preserve its length. When pulling the opposite handle at a high angle, currently, it lengthens quite a bit. If you make a handle almost vertical, the opposite handle shoots really far.

https://user-images.githubusercontent.com/12694995/135754792-4592a353-dd55-4443-b3d2-f2cae1d414c8.mp4

**Comment by fire** [comment]
For the unresolved documentation error you can run doctool.

**Comment by Razoric480** [comment]
So after having a play-around in Blender, I tried the most naive approach I could, which is just getting the vector's length and multiplying it by the original handle's length, regardless of the scale. Which gives the following copy-cat:

https://user-images.githubusercontent.com/2823241/135873398-fc380885-1ac2-4dff-8075-c3ad9bde050f.mp4

The value for the key and handles starts at the same (30s, 7 units, handles at -10/+10 seconds and 0.) Rotating the right handle causes the in-value to reach all the way up to 17 or so, and -2 or so, which is the same as it does in Godot.

**Comment by NathanLovato** [comment]
@Razoric480 I tried that already but it doesn't work. The handles get completely squashed on the other side. 

Screenshot from 2021-10-04 09-15-17

Here I have a 3-sec animation, so the X values are super small compared to the Y values, preventing us from keeping the calculations intuitive. I didn't find a way to make that work, that's why I thought we'd need to change the bezier handle values themselves.

**Comment by Razoric480** [comment]
Yeah, I see.

To be fair, the exact same issue exists in Blender if you scale the bezier editor so that it resembles Godot's scale.

https://user-images.githubusercontent.com/2823241/135881407-ccf20011-69b3-4805-b9d1-9b58943d1563.mp4

**Comment by Razoric480** [comment]
The question is if it's something to be solved. The reason Godot's graph editor starts at the size it does is based around 2D video game animation, where the amount of time per animation is short, but the units can be in hundreds of pixels of motion, whereas Blender is implying 3D where something might move a couple meters in the same span of time.

As seen above, if you decrease blender's time and increase its units to resemble Godot, you end up with the same problem where you have a handle that's 0.1 units long.

I struggle to think of anything to replace it with and have it still relate to the editor it's in. If the user changes the scale of the seconds or Y axis units, then the points should change so that the interpolation looks the same - so they remain tied to seconds and Y unit scale even then. So I don't think we can change them from seconds/units and have them relate to zooming.

We could start the bezier editor more square, either increasing the length of the seconds or shortening the scale of the Y, but then the user zooms out and the problem remains.

I suppose we have to wonder how much of an issue it really is. The user can zoom back in and change them then. Alternatively, we could have a 'ghost' handle that keeps the bezier handle a certain radius from the keyframe, even if the handle itself is very short.

**Comment by NathanLovato** [comment]
It's a usability issue if with the animations you do most commonly (less that a second to a couple seconds) lead to bezier handles jumping around strangely.

Right now it's definitely off, even the interpolation as illustrated on the screenshot I posted above: by barely moving a handle up, instead of having the curve smoothly mirror on the left side, it turned into a sharp upward slope.

In Blender, while you might be able to artificially re-create the issue, creating real game animations in the past, I've not experienced anything like it. If we could get to some compromise that makes the user experience smooth in almost 100% of cases like Blender, that'd be fine, I think.

And so it seems this is a challenge mostly for tracks that lead to large values, like pixels in 2D, or any property that would use super large values (say, animating the camera limits?), right? Any ideas on how to make that work?

**Comment by Razoric480** [comment]
We don't exactly _care_ all that much about the internal value of the in/out handles. So long as the values and seconds work for the interpolation calculation that `Animation::bezier_track_interpolate()` does, then that's fine.

We just need it to be more usable in the editor, even when it's (0.05, 200) or whatever. So we can just draw/detect input as if it was longer, even if internally it's tiny.

https://user-images.githubusercontent.com/2823241/135930213-b7c31e6b-e054-4321-a1e4-37719d892087.mp4

I just hard-coded a minimum of 30 pixels here, but would be some scalar based on the x and y axes zoom.

**Comment by NathanLovato** [comment]
The problem, in that case, is the interpolation of the curve, you can see how turning the handle produces sharp lines on the opposite side. You expect and need a smooth, continuous curve instead.

You can see on your video how the controls are still janky, with the curve jumping as you turn the handle. Blender's a good reference for how it should work out of the box for the user.

**Comment by Razoric480** [comment]
Really struggling to think of a solution after poking at it.

If we use Blender as an example, they suffer the same issue if you make their graph editor's scale the same as what you'd find in Godot 2D animation. Mathematically correct, but makes for poor curves.

Inkscape feels good, but it does not allow you to zoom in more horizontally/vertically than the other. The X and the Y units are the same.

Autodesk Maya seems to get it perfect regardless of zoom level.

https://user-images.githubusercontent.com/2823241/136103666-6923fb4b-0884-47a1-81ae-d698c5ed1e99.mp4

Though what it's doing to come up with its math is anyone's guess. The MEL commands and the [MFnAnimCurve class](https://help.autodesk.com/view/MAYAUL/2016/ENU/?guid=__cpp_ref_class_m_fn_anim_curve_html) provide some detail about how they are stored and calculated based on commands, but not how they are mirrored so nicely.

**Comment by NathanLovato** [comment]
> which feels like it defeats the purpose of having the information on the key.

Having it on the key at least makes it so you don't risk inadvertently changing the mode by clicking on a handle (that's how it is without this PR). It also allows the user to change the mode on all selected keys at once (and cause the handles to balance for instance).

I'm curious as to how maya and other programs solve that (e.g. after effects must have to deal with the same issue).

**Comment by NathanLovato** [comment]
Makes sense. Well, I guess we can move back the calculation to the bezier editor where we can. Perhaps not expose the properties to the inspector and have the user always go through the editor? Now that I think of it, for example in Blender, you can only change keyframe handle type from the animation graph editor.

**Comment by Razoric480** [comment]
You can set the handle values themselves in Blender. 
image

We'd probably want to maintain the current value-types because the interpolation uses them as they currently are. And if we're going to have the values as-is anyway, then we can leave them exposed.

**Comment by NathanLovato** [comment]
> You can set the handle values themselves in Blender.

Sure but that's from within their bezier curve editor so they could probably access the editor's scale, which they're missing. And well they don't have the same issues we're having here perhaps.

**Comment by Razoric480** [comment]
https://user-images.githubusercontent.com/2823241/136219809-a8f76478-8714-4cfd-b6b6-90c23a2053ef.mp4

**Comment by NathanLovato** [comment]
Looking good!

**Comment by Razoric480** [comment]
Leaves one final decision, I think:

1. Editing the handle in the inspector causes the handle mode to go from Balanced to Free
2. Editing the handle in the inspector leaves the handle mode Balanced, even though they're no longer flat. But they'll snap back into place once the user drags either handle in the graph editor.
3. Remove the handles from the inspector to prevent manual editing

**Comment by Razoric480** [comment]
Added that signal. Only issue is that I couldn't really reconcile the change with the undo/redo system without creating an infinite cycle, so while you can undo from balanced back to free, it won't reset the position of the handle that got balanced.

That's inline with how it works for something like, I.E., the gradient editor, where you can't undo/redo moving the gradients' points, but should be investigated.

**Comment by fire** [comment]
I tried the bezier handles in a build and it's neat. I like it too.

**Comment by fire** [comment]
@NathanLovato Do you have any ideas how to turn a regular track into bezier view. Wanted some ideas in a proposal.

**Comment by golddotasksquestions** [comment]
> 
> howaboudat.mp4

I really love how you guys try to work on and fix these animation curves, but I have to say this does not look good to me at all. With Bezier curves the handles are always supposed to represent to amplitude/strength as well as the direction of the manipulation. This seems to behave in a totally different way.

**Comment by NathanLovato** [comment]
@golddotasksquestions this is an outdated video clip, it was just a test by François. It works the same as in Blender or other programs now.

**Review by groud — COMMENTED** [review]
As is, the implementation looks correct, but i wonder if this is the simplest approach.

Wouldn't it be simpler to modify the `bezier_track_set_key_in_handle` and `bezier_track_set_key_out_handle` to update the previous/next handle if the mode is balanced ? This approach would mean that the API would consider the handle mode too.

**Comment by Razoric480** [comment]
> Edit: Oh my bad I misunderstood the issue. You want to keep the length on the other side! I understand the issue better now.

Was about to say, yeah, that was the main issue with flipping the handle around. :) The alternative is to provide the Animation resource with some largely unrelated data about the scale of the UI and that feels like the wrong approach.

**Comment by groud** [comment]
> Was about to say, yeah, that was the main issue with flipping the handle around. :) The alternative is to provide the Animation resource with some largely unrelated data about the scale of the UI and that feels like the wrong approach.

I am not really sure. I think the only think needed to be provided to the Animation resource would be the ratio between the Y-axis scale and the X-axis scale. So a simple float value. It might not be a big of a problem to add this as an optional argument to the `bezier_track_set_key_in/out_handle` functions I think.

My main problem with the implementation right now is that we are adding a property to each handle in the Animation resource, while this is something that will only stay usable from the editor. We usually try to avoid this and store the information elsewhere, not to fill the core classes with editor-only stuff. 

So, IMHO, we should decide what kind of approach we want:
- If we go full-editor version, then we should probably remove the `handle_mode` properties from handles. We would then implement an EditorInspectorPlugin that would add a field in the AnimationTrackHandle inspector. The value itself could then maybe be stored in the metadata dict of the handle (prefixed by "_edit_") as this is a trick we used often in the codebase (we may have to change that someday to a dedicated editor_metadata dictionary, but we have to decide later on).
- We go full API-version, and then we add everything needed to the API to be able to handle the balanced mode correctly. This means adding the "YXscale_ratio" to the `bezier_track_set_key_in/out_handle` functions as I mentioned above. 
 
I think the later is simpler to implement even if the argument is a little bit weird. I guess it can still be explained in the documentation. And even then, I am not sure there's a lot of people procedurally generating Animations like that anyway, so it can stay a little bit cryptic. ^^

**Comment by NathanLovato** [comment]
Okay, I didn't know you would use some metadata in those cases. Here, other properties like the handle vectors themselves are already exposed, which is why we went this route. If it's fine we can just keep it simple and add this argument. Like you said, I don't think too many persons would use it, and I don't think too many would use or have used the current properties of those bezier keyframes.

The only case would be to make a custom in-game animation editor at runtime as there's no plugin support for the animation editor I think.

**Comment by NathanLovato** [comment]
Okay, I didn't know you would use some metadata in those cases. Here, other properties like the handle vectors themselves are already exposed, which is why we went this route. If it's fine we can just keep it simple and add this argument. Like you said, I don't think too many persons would use it, and I don't think too many would use or have used the current properties of those bezier keyframes.

The only case would be to make a custom in-game animation editor at runtime as there's no plugin support for the animation editor I think.

**Review by groud — APPROVED** [review]
Looks good to me! Good job!

File: scene/resources/animation.h [hunk 1]
```diff
@@ -419,11 +424,13 @@ class Animation : public Resource {
...
 	InterpolationType track_get_interpolation_type(int p_track) const;
 
-	int bezier_track_insert_key(int p_track, double p_time, real_t p_value, const Vector2 &p_in_handle, const Vector2 &p_out_handle);
+	int bezier_track_insert_key(int p_track, double p_time, real_t p_value, const Vector2 &p_in_handle, const Vector2 &p_out_handle, const HandleMode p_handle_mode = HandleMode::HANDLE_MODE_BALANCED);
+	void bezier_track_set_key_handle_mode(int p_track, int p_index, HandleMode p_mode, const double &p_balanced_value_time_ratio = 1.0);
 	void bezier_track_set_key_value(int p_track, int p_index, real_t p_value);
-	void bezier_track_set_key_in_handle(int p_track, int p_index, const Vector2 &p_handle);
-	void bezier_track_set_key_out_handle(int p_track, int p_index, const Vector2 &p_handle);
+	void bezier_track_set_key_in_handle(int p_track, int p_index, const Vector2 &p_handle, const double &p_balanced_value_time_ratio = 1.0);
+	void bezier_track_set_key_out_handle(int p_track, int p_index, const Vector2 &p_handle, const double &p_balanced_value_time_ratio = 1.0);
```
Comment (akien-mga) [inline scene/resources/animation.h:431]: `const double &` seems a bit overkill for a primitive type like this, `double` should be enough.

Comment (Razoric480) [reply scene/resources/animation.h]: Fixed, thanks.

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks a lot to all involved!

**Pull request merged**